compiler/backend/src/org/jetbrains/kotlin/codegen/ImplementationBodyCodegen.java

```diff
@@ -574,8 +575,20 @@ else if (asmType.getSort() == Type.DOUBLE) {
                     iv.ifne(ne);
                 }
                 else {
-                    StackValue value = genEqualsForExpressionsOnStack(KtTokens.EQEQ, StackValue.onStack(asmType), StackValue.onStack(asmType));
-                    value.put(Type.BOOLEAN_TYPE, iv);
+                    if (isPrimitive(asmType)) {
+                        StackValue value = StackValue.cmp(KtTokens.EQEQ, asmType, StackValue.onStack(asmType), StackValue.onStack(asmType));
+                        value.put(Type.BOOLEAN_TYPE, iv);
+                    } else {
+                        if (DescriptorUtils.isInterface(propertyDescriptor.getType().getConstructor().getDeclarationDescriptor())) {
+                            iv.invokeinterface(asmType.getInternalName(), "equals",
```

Comment:
I think it's fine to continue calling `invokevirtual java/lang/Object.equals` for interfaces, since it's not immediately clear that `invokeinterface I.equals` is better. So I propose to replace this if with:
```
Type owner =
        DescriptorUtils.isInterface(propertyDescriptor.getType().getConstructor().getDeclarationDescriptor())
        ? AsmTypes.OBJECT_TYPE
        : asmType;
iv.invokevirtual(
        owner.getInternalName(), "equals", Type.getMethodDescriptor(Type.BOOLEAN_TYPE, AsmTypes.OBJECT_TYPE), false
);
```